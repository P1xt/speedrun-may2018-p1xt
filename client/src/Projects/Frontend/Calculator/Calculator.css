.Calculator {
  text-align: left;
  margin-top: 40px; }
  .Calculator > h1 {
    text-transform: none; }
