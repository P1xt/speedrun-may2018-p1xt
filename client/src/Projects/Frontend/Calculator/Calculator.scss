.Calculator {
  text-align: center;
  width: 290px;
  margin: 40px auto;

  /**
   * Drop everything in a card
   */
  .card {
    background: #222222;
    color: #ffffff;
    border-radius: 50px;
    h1 {
      text-transform: none;
      color: #ffffff;
      margin-top: 20px;
    }
    h2 {
      color: #444444;
    }
  }
  
  /**
   * Overall design of everything in the card
   */
  &_contents {
    font-size: 24px;
    margin: 20px auto;
    width: 225px;
  }

  /**
   * input and results display
   */
  &_display {
    color: #000000;
    background-color: #cccccc;
    margin-top: 5px;
    margin-bottom: 5px;
    margin-left: 5px;
    width: 155px;
    height: 45px;
    float: left;
    line-height: 45px;
    text-align: right;
    overflow:hidden !important;
  }  

  /**
   * Design of the individual buttons
   */
  button {
    margin: 5px;
    width: 45px;
    height: 45px;
    border-radius: 5px;
    border: 0;
    background: #757575;
    color: #ffffff;
    cursor: pointer;
    float: left;
    &:hover {
      background: #333333;
    }
  }  

}
