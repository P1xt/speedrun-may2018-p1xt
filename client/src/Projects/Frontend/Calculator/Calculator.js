import React, { Component } from "react";
import { Card } from "reactstrap";

import "./Calculator.css";
/**
 * Implements a simple calculator component
 * that provides addition, subtraction,
 * multiplication and division of integers
 * and floating point numbers
 * @class Calculator
 * @extends Component
 */
class Calculator extends Component {
  constructor() {
    super();

    this.state = {
      display: "",
      warning: ""
    };
  }

  /**
   * processInput
   * Appends input to the current calculation
   * if the input is valud.
   *
   * Invalid possibilities:
   * - two decimal points in a row
   * - an operator at the beginning before a number
   * - more characters than the display can hold (10)
   * - two operators in a row (++)
   * @memberof Calculator
   * @param input: string
   */
  processInput = input => {
    const displayLast = this.state.display.slice(-1);
    const operators = {
      "+": 0,
      "-": 0,
      "*": 1,
      "/": 1
    };
    if (this.state.display.length > 9) {
      this.setState({
        warning: "Input full"
      });
    } else if (
      (displayLast === "." && input === ".") ||
      (operators.hasOwnProperty(displayLast) &&
        operators.hasOwnProperty(input) &&
        input !== "-") ||
      (displayLast === "" && operators.hasOwnProperty(input))
    ) {
      this.setState({
        warning: "Invalid Input"
      });
    } else {
      this.setState({
        warning: "",
        display: this.state.display + input
      });
    }
  };

  /**
   * processClear
   * Clear the current calculation
   * @memberof Calculator
   */
  processClear = input => {
    this.setState({
      warning: "",
      display: ""
    });
  };

  /**
   * processResult
   * Evaluate the current calculation
   *
   * Note: uses eval which is both unsafe
   * and lazy - but this is for a speedrun
   * and reverse polish eval from scratch
   * would take a few hours and a few hundred
   * lines of code I don't have time for right now
   * @memberof Calculator
   */
  processResult = () => {
    // eslint-disable-next-line
    const result = eval(this.state.display);

    if (result) {
      this.setState({
        warning: "",
        display: "" + result
      });
    }
  };

  /**
   * render
   * Renders the current component
   * @return jsx
   * @memberof Calculator
   */
  render() {
    return (
      <div className="Calculator">
        <Card>
          <h1>JavaScript Calculator</h1>
          <h2>{this.state.warning}</h2>
          <div className="Calculator_contents">
            <button
              onClick={() => {
                this.processClear();
              }}
            >
              C
            </button>
            <div className="Calculator_display">{this.state.display}</div>
            <button
              onClick={() => {
                this.processInput("7");
              }}
            >
              7
            </button>
            <button
              onClick={() => {
                this.processInput("8");
              }}
            >
              8
            </button>
            <button
              onClick={() => {
                this.processInput("9");
              }}
            >
              9
            </button>
            <button
              onClick={() => {
                this.processInput("+");
              }}
            >
              +
            </button>
            <button
              onClick={() => {
                this.processInput("4");
              }}
            >
              4
            </button>
            <button
              onClick={() => {
                this.processInput("5");
              }}
            >
              5
            </button>
            <button
              onClick={() => {
                this.processInput("6");
              }}
            >
              6
            </button>
            <button
              onClick={() => {
                this.processInput("-");
              }}
            >
              -
            </button>
            <button
              onClick={() => {
                this.processInput("1");
              }}
            >
              1
            </button>
            <button
              onClick={() => {
                this.processInput("2");
              }}
            >
              2
            </button>
            <button
              onClick={() => {
                this.processInput("3");
              }}
            >
              3
            </button>
            <button
              onClick={() => {
                this.processInput("*");
              }}
            >
              *
            </button>
            <button
              onClick={() => {
                this.processInput("0");
              }}
            >
              0
            </button>
            <button
              onClick={() => {
                this.processInput(".");
              }}
            >
              .
            </button>
            <button
              onClick={() => {
                this.processResult();
              }}
            >
              =
            </button>
            <button
              onClick={() => {
                this.processInput("/");
              }}
            >
              /
            </button>
          </div>
        </Card>
      </div>
    );
  }
}

export default Calculator;
