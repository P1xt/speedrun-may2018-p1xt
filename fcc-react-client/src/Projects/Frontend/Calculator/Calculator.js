import React, { Component } from "react";
import { Card } from "reactstrap";

import "./Calculator.css";

/**
 * Implements a simple calculator component
 * that provides addition, subtraction,
 * multiplication and division of integers
 * and floating point numbers
 * @class Calculator
 * @extends Component
 */
class Calculator extends Component {
  constructor() {
    super();

    this.state = {
      display: "",
      warning: ""
    };
  }
  /**
   * processEntry
   * Handles user input and passes it to the
   * appropriate function
   * @memberof Calculator
   */
  processEntry = e => {
    const input = e.target.value;
    if (input === "=") {
      this.processResult();
    } else {
      this.processInput(input);
    }
  };

  /**
   * processInput
   * Appends input to the current calculation
   * if the input is valud.
   *
   * Invalid possibilities:
   * - two decimal points in a row
   * - an operator at the beginning before a number
   * - more characters than the display can hold (10)
   * - two operators in a row (++)
   * @memberof Calculator
   * @param input: string
   */
  processInput = input => {
    const displayLast = this.state.display.slice(-1);
    const operators = ["+", "-", "*", "/"];

    if (this.state.display.length > 9) {
      this.setState({
        warning: "Input full"
      });
    } else if (
      (displayLast === "." && input === ".") ||
      (operators.includes(displayLast) && operators.includes(input)) ||
      (displayLast === "" && operators.includes(input))
    ) {
      this.setState({
        warning: "Invalid Input"
      });
    } else {
      this.setState({
        warning: "",
        display: this.state.display + input
      });
    }
  };

  /**
   * processClear
   * Clear the current calculation
   * @memberof Calculator
   */
  processClear = input => {
    this.setState({
      warning: "",
      display: ""
    });
  };

  /**
   * processResult
   * Evaluate the current calculation
   *
   * Note: uses eval which is both unsafe
   * and lazy - but this is for a speedrun
   * and reverse polish eval from scratch
   * would take a few hours and a few hundred
   * lines of code I don't have time for right now
   * @memberof Calculator
   */
  processResult = () => {
    const displayLast = this.state.display.slice(-1);
    const operators = ["+", "-", "*", "/"];

    // catch trailing operator so it doesn't break eval
    if (operators.includes(displayLast)) {
      this.setState({
        warning: "Invalid Input"
      });
    } else {
      // eslint-disable-next-line
      const result = eval(this.state.display);
      if (result) {
        this.setState({
          warning: "",
          display: "" + result
        });
      }
    }
  };

  /**
   * render
   * Renders the current component
   * @return jsx
   * @memberof Calculator
   */
  render() {
    return (
      <div className="Calculator">
        <Card>
          <h1>JavaScript Calculator</h1>
          <h2>{this.state.warning}</h2>
          <div className="Calculator_contents">
            <button onClick={this.processClear}>C</button>
            <div className="Calculator_display">{this.state.display}</div>
            {[
              "7",
              "8",
              "9",
              "+",
              "4",
              "5",
              "6",
              "-",
              "1",
              "2",
              "3",
              "*",
              "0",
              ".",
              "=",
              "/"
            ].map(button => (
              <button key={button} value={button} onClick={this.processEntry}>
                {button}
              </button>
            ))}
          </div>
        </Card>
      </div>
    );
  }
}

export default Calculator;
