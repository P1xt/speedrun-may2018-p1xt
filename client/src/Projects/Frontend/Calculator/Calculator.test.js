import React from "react";
import renderer from 'react-test-renderer';
import { shallow } from "enzyme";

import Calculator from "./Calculator";

describe("<Calculator />", () => {
  it("renders without crashing", () => {
    shallow(<Calculator />);
  });
  it("renders correctly", () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
