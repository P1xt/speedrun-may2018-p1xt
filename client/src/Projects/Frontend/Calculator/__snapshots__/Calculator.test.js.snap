// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator /> renders correctly 1`] = `
<div
  className="Calculator"
>
  <div
    className="card"
  >
    <h1>
      JavaScript Calculator
    </h1>
    <h2>
      
    </h2>
    <div
      className="Calculator_contents"
    >
      <button
        onClick={[Function]}
      >
        C
      </button>
      <div
        className="Calculator_display"
      >
        
      </div>
      <button
        onClick={[Function]}
        value="7"
      >
        7
      </button>
      <button
        onClick={[Function]}
        value="8"
      >
        8
      </button>
      <button
        onClick={[Function]}
        value="9"
      >
        9
      </button>
      <button
        onClick={[Function]}
        value="+"
      >
        +
      </button>
      <button
        onClick={[Function]}
        value="4"
      >
        4
      </button>
      <button
        onClick={[Function]}
        value="5"
      >
        5
      </button>
      <button
        onClick={[Function]}
        value="6"
      >
        6
      </button>
      <button
        onClick={[Function]}
        value="-"
      >
        -
      </button>
      <button
        onClick={[Function]}
        value="1"
      >
        1
      </button>
      <button
        onClick={[Function]}
        value="2"
      >
        2
      </button>
      <button
        onClick={[Function]}
        value="3"
      >
        3
      </button>
      <button
        onClick={[Function]}
        value="*"
      >
        *
      </button>
      <button
        onClick={[Function]}
        value="0"
      >
        0
      </button>
      <button
        onClick={[Function]}
        value="."
      >
        .
      </button>
      <button
        onClick={[Function]}
        value="="
      >
        =
      </button>
      <button
        onClick={[Function]}
        value="/"
      >
        /
      </button>
    </div>
  </div>
</div>
`;
